### PR TITLE
make IO::Uncompress::Gunzip optional

### PR DESCRIPTION
**Pull request by skaji**
Address #16 again.

This PR makes IO::Uncompress::Gunzip optional.
If it is not available, CPAN::Common::Index::Mirror will fetch uncompressed index files,
and CPAN::Common::Index::LocalPackage will accept uncompressed index files only.

off topic:
I added t/CPAN/modules/02packages.details.txt for tests,
which makes repo +11MB, and CPAN tarball +1.5MB.
Actually we don't need whole 02pacakges.details.txt, so I could add a shrinked version of 02packages.details.txt.

### CODE REVIEW

**Comment by xdg** [comment]
Yes, please do shrink the uncompressed files.

I think there should probably be tests that hide IO::Uncompress::Gunzip and verify that the behaviors are what we expect.

**Comment by skaji** [comment]
@xdg Thank you for your comment.

I have shrinked 02packages.details.txt and 01mailrc.txt in b4fc15a (force pushed).
And tests run with hiding IO::Uncompress::Gunzip too.

**Comment by skaji** [comment]
@xdg Could you check this agian please?

**Comment by xdg** [comment]
I'm on vacation and will be busy this week when I return.  If I don't review by next week, please feel free to remind me.

**Comment by skaji** [comment]
@xdg ping?

File: t/local_package.t [hunk 1]
```diff
@@ -67,44 +68,66 @@ subtest "constructor tests" => sub {
...
 
 # XXX test that files in cache aren't overwritten?
 
-subtest 'check index age' => sub {
-    my $index   = new_local_index;
-    my $package = $index->cached_package;
-    ok( -f $package, "got the package file" );
-    my $expected_age = ( stat($package) )[9];
-    is( $index->index_age, $expected_age, "index_age() is correct" );
+my $common_tests = sub {
```
Comment (xdg) [inline t/local_package.t]: I don't see the benefit of making this an anonymous subroutine.  Doing so makes me search for some reason why it's not a normal subroutine, so I'd prefer to have it just be a  normal sub (unless there is some reason I'm missing).

File: t/local_package.t [hunk 2]
```diff
@@ -67,44 +68,66 @@ subtest "constructor tests" => sub {
 };
 
 subtest 'refresh and unpack index files' => sub {
+    plan skip_all => "IO::Uncompress::Gunzip is not available"
+      unless $CPAN::Common::Index::Mirror::HAS_IO_UNCOMPRESS_GUNZIP;
     my $index = new_local_index;
 
-    ok( !-e File::Spec->catfile( $cache, $packages ), "$packages not in cache" );
+    ok( !-e File::Spec->catfile( $index->cache, $packages ), "$packages not in cache" );
```
Comment (xdg) [inline t/local_package.t]: This is a subtle semantic change in the test here and on lines 79, 85, and 90.  Originally, the test is making sure that the file is/isn't in the directory we passed to the constructor.  The new test could completely ignore the cache attribute in the constructor as long as the cache *method* returns a directory where things actually do/don't live.  So we know longer have a assurance that the cache constructor attribute is respected.

You could address that in `new_local_index` and `new_uncompressed_local_index` by testing that `$index->cache eq $cache` before returning `$index`.

File: t/mirror.t [hunk 3]
```diff
@@ -15,11 +15,11 @@ use CommonTests;
 
 my $cwd         = getcwd;
 my $test_mirror = "file:///$cwd/t/CPAN";
-my $cache       = File::Temp->newdir;
 my $mailrc      = "01mailrc.txt";
 my $packages    = "02packages.details.txt";
 
 sub new_mirror_index {
+    my $cache = File::Temp->newdir;
```
Comment (xdg) [inline t/mirror.t]: As in the last test file, you should now check that after `new_ok`, that `$index->cache eq $cache`.

File: t/mirror.t [hunk 4]
```diff
@@ -55,44 +55,60 @@ subtest "constructor tests" => sub {
...
-
-subtest 'find package' => sub {
-    my $index = new_mirror_index;
-    test_find_package($index);
-};
-
-subtest 'search package' => sub {
-    my $index = new_mirror_index;
-    test_search_package($index);
+my $common_tests = sub {
```
Comment (xdg) [inline t/mirror.t]: As before, make a regular subroutine unless there's a reason not to do so.

**Comment by skaji** [comment]
@xdg Thank you for your review!

* df23c77 Because there is no reason to use anonymous subs, change them to normal subs.
* be4dc32 You're right, I missed an assurance that the cache constructor attribute was respected. Now check it in new_XXX_index() methods.

**Comment by skaji** [comment]
@xdg thanks!

**Comment by skaji** [comment]
@xdg Do you have a plan to release a stable version?

**Pull request closed without merging**